go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -391,3 +668,526 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+func (vbc *vcopierCopyWorker) IsSameSetting(string) bool {
+	return true
+}
+
+// IsSettingApplied implements pools.Resource.
+func (vbc *vcopierCopyWorker) IsSettingApplied() bool {
+	return false
+}
+
+// ResetSetting implements pools.Resource.
+func (vbc *vcopierCopyWorker) ResetSetting(context.Context) error {
+	return nil
+}
+
+// execute advances a task through each state until it is canceled, completed
+// or failed.
+func (vbc *vcopierCopyWorker) execute(ctx context.Context, task *vcopierCopyTask) *vcopierCopyTaskResult {
+	startedAt := time.Now()
+	state := vcopierCopyTaskPending
+
+	var err error
+
+	// As long as the current state is not done, keep trying to advance to the
+	// next state.
+	for !vcopierCopyTaskStateIsDone(state) {
+		// Get the next state that we want to advance to.
+		nextState := vcopierCopyTaskGetNextState(state)
+
+		var advanceFn func(context.Context, *vcopierCopyTaskArgs) error
+
+		// Get the advanceFn to use to advance the task to the nextState.
+		switch nextState {
+		case vcopierCopyTaskBegin:
+			advanceFn = func(context.Context, *vcopierCopyTaskArgs) error {
+				// Rollback to make sure we're in a clean state.
+				if err := vbc.vdbClient.Rollback(); err != nil {
+					return fmt.Errorf("failed to rollback: %s", err.Error())
+				}
+				// Begin transaction.
+				if err := vbc.vdbClient.Begin(); err != nil {
+					return fmt.Errorf("failed to start transaction: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskInsertRows:
+			advanceFn = func(ctx context.Context, args *vcopierCopyTaskArgs) error {
+				if _, err := vbc.insertRows(ctx, args.rows); err != nil {
+					return fmt.Errorf("failed inserting rows: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskInsertCopyState:
+			advanceFn = func(ctx context.Context, args *vcopierCopyTaskArgs) error {
+				if err := vbc.insertCopyState(ctx, args.lastpk); err != nil {
+					return fmt.Errorf("error updating _vt.copy_state: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskCommit:
+			advanceFn = func(context.Context, *vcopierCopyTaskArgs) error {
+				// Commit.
+				if err := vbc.vdbClient.Commit(); err != nil {
+					return fmt.Errorf("error commiting transaction: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskComplete:
+			advanceFn = func(context.Context, *vcopierCopyTaskArgs) error { return nil }
+		default:
+			err = fmt.Errorf("don't know how to advance from %s to %s", state, nextState)
+			state = vcopierCopyTaskFail
+			goto END
+		}
+
+		// tryAdvance tries to execute advanceFn, and also executes any
+		// lifecycle hooks surrounding the provided state.
+		//
+		// If all lifecycle hooks and advanceFn are successfully executed,
+		// tryAdvance will return state, which should be == the nextState that
+		// we provided.
+		//
+		// If there was a failure executing lifecycle hooks or advanceFn,
+		// tryAdvance will probably return a failure state (i.e. canceled or
```

Comment:
`probably` makes me a little nervous here 🙂 We should have some code to handle the unexpected/edge cases